crates/goose-server/src/routes/reply.rs

```diff
@@ -294,18 +299,40 @@ async fn handler(
             }
         };
 
-        while let Some(response) = stream.next().await {
-            match response {
-                Ok(message) => {
-                    if let Err(e) = stream_message(message, &tx).await {
-                        tracing::error!("Error sending message through channel: {}", e);
-                        break;
+        loop {
+            tokio::select! {
+                response = timeout(Duration::from_millis(500), stream.next()) => {
+                    match response {
+                        Ok(Some(Ok(message))) => {
+                            if let Err(e) = stream_message(message, &tx).await {
+                                tracing::error!("Error sending message through channel: {}", e);
+                                break;
+                            }
+                        }
+                        Ok(Some(Err(e))) => {
+                            tracing::error!("Error processing message: {}", e);
+                            break;
+                        }
+                        Ok(None) => {
+                            break;
+                        }
+                        Err(_) => { // Heartbeat, used to detect disconnected clients and then end running tools.
+                            if let Err(e) = tx.try_send(ProtocolFormatter::heartbeat()) {
```

Comment:
nit: should we use `tx.is_closed` instead of sending a heartbeat?

Reply:
That worked somehow, I've removed the sending of heartbeat events.